sharktank/sharktank/layers/mixture_of_experts_block.py

```diff
@@ -0,0 +1,115 @@
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from typing import Optional
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+from .base import Theta, ThetaLayer
+from .linear import LinearLayer
+from .norm import RMSNormLayer
+from .ffn_moe_block import FFNMOE
+
+__all__ = [
+    "SparseMoeBlock",
+]
+
+
+class SparseMoeBlock(ThetaLayer):
+    """
+    This implementation considers MoE operations as block-sparse
+    operations to support imbalanced token assignments to experts.
+    This enables the MoE to operate at a faster rate and in full capacity without any dropped tokens
+    (or reduced performance).
+    """
+
+    def __init__(
+        self,
+        theta: Theta,
+        expert_count: int,
+        expert_used_count: int,
+        rms_epsilon: float,
+    ):
+        super().__init__(theta)
+
+        # Add router gate
+        self.add_module("ffn_gate_inp", LinearLayer(theta("ffn_gate_inp")))
+
+        # Add FFN norm
+        self.add_module(
+            "ffn_norm", RMSNormLayer(theta("ffn_norm"), epsilon=rms_epsilon)
+        )
+
+        # Add FFN output norm
+        self.add_module(
+            "layer_output_norm",
+            RMSNormLayer(theta("layer_output_norm"), epsilon=rms_epsilon),
+        )
+
+        # Add expert_count x FFN
+        self.experts = nn.ModuleList(
+            [FFNMOE(theta, expert_idx=i) for i in range(expert_count)]
+        )
+
+        self.expert_count = expert_count
+        self.expert_used_count = expert_used_count
+
+    def forward(
+        self,
+        h: torch.Tensor,
+    ):
+        ffn_input = self.ffn_norm(h)
+        batch_size, sequence_length, feature_dim = ffn_input.shape
+        ffn_input = ffn_input.view(-1, feature_dim)
+
+        # For each token, the router calculates the router weights for all experts
+        # router_logits: (batch_size * sequence_length, expert_count)
+        router_logits = self.ffn_gate_inp(ffn_input)
+        router_weights = F.softmax(router_logits, dim=1, dtype=torch.float)
+
+        # Select top k experts from router weights
+        router_weights, top_k_experts = torch.topk(
+            router_weights, self.expert_used_count, dim=-1
+        )
+        router_weights /= router_weights.sum(dim=-1, keepdim=True)
+        router_weights = router_weights.to(ffn_input.dtype)
+
+        moe_output = torch.zeros(
+            (batch_size * sequence_length, feature_dim), dtype=ffn_input.dtype
+        )
+
+        # Create an expert mask by one hot encoding the selected top k experts
+        # used to index which expert is to be invoked for each token
+        # expert_mask: (expert_count, expert_used_count, sequence_length)
+        expert_mask = F.one_hot(top_k_experts, num_classes=self.expert_count).permute(
+            2, 1, 0
+        )
+
+        # Iterate over all experts in the model
+        for expert_idx in range(self.expert_count):
```

Comment:
Are we replacing this with slow_matmul for Grok?

Reply:
I'll add this in a follow up patch that replaces the expert splitting in the experts too